Add repr() printing & dummy=False for DifferentialExtension object

gxyd opened the pull request:
<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->

gxyd committed:
Add repr() printing of DifferentialExtension object

asmeurer commented:
Are you planning to add more here? I think it would be helpful if the str form indicated what the different elements were somehow. 

You might also want repr to use the `extension` argument instead of depending on `f` and `x`, as some extensions can't be generated from a given `f` yet (e.g., to create a tangent extension currently, you have to do it manually, because the algorithms aren't there yet to process tangents).

gxyd committed:
change dummy=True to dummy=False in DifferentialExtension

We only need "dummy=True" in the algorithm itself. When testing
interactively we want "dummy=False". And also in the tests
so "dummy=False" would probably be more useful as a default

gxyd commented:
I have addressed the change `dummy=True` -> `dummy=False` and `str` printing of `DifferentialExtension` object. I don't intend to add any more changes in this PR.

gxyd committed:
remove un-necessary `dummy=False`, since that is default

gxyd commented:
Should I add `dummy=True` for the tests that had initially `dummy=True` by default, to make the tests behave the way they did earlier? Though I have removed the un-necessary `dummy=False` from the ones using it.

gxyd commented:
@asmeurer please review.

asmeurer commented:
I think unless the test is explicitly testing the dummy parameter, we shouldn't bother testing dummy=True.

gxyd commented:
I don't think that any test is there to explicitly test the condition of `dummy=True`.

asmeurer commented:
You need to set dummy=True where the DifferentialExtension object is actually used in the algorithm. A test to add for this would be an expression that has constants named `t0`, to make sure it doesn't confuse the algorithm.

asmeurer commented:
That is, something like `risch_integrate(t0*exp(x), x)`.

jksuom commented:
dummy=True could probably be set internally, in risch_integrate. That way it would be easier to do debugging with default dummy=False.

gxyd committed:
use dummy=True as default in `risch_integrate`

asmeurer commented:
This looks good now.

gxyd commented:
Actually I was just confusing earlier about your comment to changing it in "the algorithm"(I should have understood though), but Kalevi's explicit comment made it crystal clear. :)

gxyd commented:
Unless this requires any changes, can we please merge this (failing test is unrelated)? Since I want to make use of its commits.

jksuom commented:
We may have to rerun the whole sequence of tests. See https://www.traviscistatus.com/incidents/93wfcv0ksqrz

jksuom commented on sympy/integrals/risch.py:
```diff
@@ -1668,7 +1673,8 @@ def risch_integrate(f, x, extension=None, handle_first='log',
     f = S(f)
 
-    DE = extension or DifferentialExtension(f, x, handle_first=handle_first, rewrite_complex=rewrite_complex)
+    DE = extension or DifferentialExtension(f, x, handle_first=handle_first, dummy=True, \
```
Backslash should not be necessary here in the argument list.

jksuom commented:
So the restart of https://travis-ci.org/sympy/sympy/jobs/216874027 did succeed though there seemed to be no response.

But it looks like the internal methods of DifferentialExtension: _exp_part,  _log_part and reset should have dummy=dummy instead of the hard-coded dummy=False (or True).

gxyd commented:
If you mean that change in the API of the internal methods. What would the second dummy be in `dummy=dummy`? Since the second `dummy` isn't a class variable, it would raise a `NameError`.

asmeurer commented:
You could add dummy to the `__slots__` and save it. It doesn't matter to break the API (I only used `__slots__` to make it easier to keep track of all the attributes).

gxyd committed:
Add dummy to __slots__ in DifferentialExtension

gxyd commented:
I think my branch will fail tests (as it does locally), possibly because of some changes introduced in `physics` module the changes in `units` sub-folder. May be I will have to rebase. But will wait till tests show any results.

gxyd commented:
Aaron do code changes look fine?
Though I find it odd that, on my system locally tests fail (on one example of integral module):
```
gaurav@stallman:~/foss/sympy$ python3 ./bin/test sympy/integrals/tests/test_integrals.py 
============================================================= test process starts ==============================================================
executable:         /usr/bin/python3  (3.4.3-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python 
random seed:        82013443
hash randomization: on (PYTHONHASHSEED=2392343087)

sympy/integrals/tests/test_integrals.py[?]   Failed to import                                                                             [FAIL]

________________________________________________________________________________________________________________________________________________
_______________________________________ /home/gaurav/foss/sympy/sympy/integrals/tests/test_integrals.py ________________________________________
  File "/home/gaurav/foss/sympy/sympy/integrals/tests/test_integrals.py", line 11, in <module>
    from sympy.physics import units
  File "/home/gaurav/foss/sympy/sympy/physics/__init__.py", line 5, in <module>
    from . import units
ImportError: bad magic number in 'sympy.physics.units': b'\x03\xf3\r\n'

=========================================== tests finished: 0 passed, 1 exceptions, in 0.03 seconds ============================================
DO *NOT* COMMIT!
```
That happens since my system doesn't even have any files in directory sympy.physics.units (though it only contains .pyc files in it). As if I got half-baked commits. If I checkout a new branch from this one and apply rebase (with master) everything works just fine.

asmeurer commented:
When that happens try removing the `__pycache__` directory.

gxyd commented:
That works good. :)
